### DISCUSSION TITLE
Explicitly mentioning keyword arguments

### DISCUSSION THREAD

**Discussion by romanlutz** [fairlearn/fairlearn#664] [General]
For quite a while we've been leaving it up to users to guess keyword arguments like `sensitive_features`, and that recently got extended to `control_features`. I actually pointed that out in the PR:

https://github.com/fairlearn/fairlearn/pull/631#issuecomment-747123392​​​​​​​

I just realized this comment was sort of skipped (probably because these non-threaded responses easily get lost, especially when there are 100+ comments on a PR). More important than an individual method signature is the general practice, so I'd like to get on the same page with everyone. As you can tell from my comment I'm very much in favor of spelling out all the args. I just very much confirmed my own point of view by having to dig through this PR to even figure out what the argument (`control_features`) is called, rather than it being obvious. It's also not part of the API docs and neither is `sensitive_features`, e.g.,

https://fairlearn.github.io/master/api_reference/fairlearn.reductions.html#fairlearn.reductions.ExponentiatedGradient.fit

Thoughts? @riedgar-ms @hildeweerts @adrinjalali @MiroDudik

**Comment by riedgar-ms** [comment]
I would not be averse to making things explicit - one of my less favourite recent debugging memories was figuring out what was going on inside a subroutine with signature `some_routine(func, *args, **kwargs)`. I think that the counter-argument is that this way, `ExponentiatedGradient` didn't have to know what arguments the `Moment` objects needed - it just peeled off the ones it wanted, and passed everything else (i.e. `**kwargs`) on down the stack.

**Reply by romanlutz** [reply]
Thanks for the quick response! That's a fair point. From my point of view clarity & usability >> slightly cleaner code.

**Comment by adrinjalali** [comment]
I think we should limit the usage of `**kwargs` to where it's necessary, and that's usually when we need to pass things to an underlying object. But even then, like the `Moment` example, we could either accept a `Moment` object which is passed by the user and therefore the user passes one with whatever parameters they want, or we could accept a `momen_params: dict` which would be passed to the `Moment` object as `**moment_params`.

`**kwargs` is usually needed when an API design requires it, like when we know another object is going to pass us a bunch of `**fit_params` which we don't know what they are, and we should pass them along to a child estimator.
